Docs for Credentials (XLS-70d)

mDuo13 opened the pull request:
Spec: https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0070d-credentials
rippled PR: https://github.com/XRPLF/rippled/pull/5103

Updated with changes from https://github.com/XRPLF/XRPL-Standards/pull/232

sappenin commented on docs/xls-70d-on-chain-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+
+This feature set is designed for an ecosystem of parties with the following roles:
+
+- _Authorizers_ who want to limit specific actions or interactions to users who possess the right credentials.
```
```suggestion
- _Authorizers_ who want to limit specific actions or interactions to users who possess the correct credentials.
```

sappenin commented on docs/xls-70d-on-chain-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+This feature set is designed for an ecosystem of parties with the following roles:
+
+- _Authorizers_ who want to limit specific actions or interactions to users who possess the right credentials.
+- _Issuers_ who give out credentials to users who meet their criteria.
```
```suggestion
- _Issuers_ who provide credentials to users who meet their criteria.
```

sappenin commented on docs/xls-70d-on-chain-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+- _Issuers_ who give out credentials to users who meet their criteria.
+- _Users_ who obtain credentials from issuers and use those credentials to interact with authorizers.
+
+Each credential applies to a specific user's XRP Ledger account, and is attested by a specific credential issuer. The exact qualifications that a credential entails are open-ended, and defined by their issuers. They can range from trivial to serious, or broad to narrow. Some examples of things a credential could state:
```
> and is attested by a specific credential issuer

It's unclear what this part of the sentence is trying to say -- I lean toward simply removing it. In my current thinking, an issuance of a credential is subtly different from an attestation -- though perhaps I could be wrong. Are there times where a credential isn't attenting anything?

What about this?

```suggestion
Each credential applies to a specific user's XRP Ledger account. The exact qualifications of any particular credential are open-ended, and defined by each issuer. Credential meaning can range from trivial to serious, or broad to narrow. Some examples of things a credential could state:
```

mDuo13 replied:
I disagree on this. A credential is a signed statement by the issuer, which is IMO exactly what an _attestation_ is: a signed statement by a known party.

sappenin replied:
Good point -- re-reading this with fresh eyes, I agree with you and prefer the original wording. I'll resolve this one.

sappenin commented on docs/xls-70d-on-chain-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+
+The typical usage for Credentials features three parties with different roles. Consider the following example:
+
+* Verity is a regulated business that wants to interact only with properly KYC'd accounts, to ensure legal compliance. This makes Verity an _authorizer_ because they configure which accounts are allowed (authorized) to interact with them.
```
Do we want to use a fake biz name here instead?

mDuo13 replied:
I just reused what was in the spec, but is Verity not a fake biz name? Sounds like one to me.

mvadari replied:
It was originally a fake person name, actually. Verity the Verifier.

sappenin reviewed (commented):
Only did a review of `index.md`, FWIW.

oleks-rip commented on docs/xls-70d-on-chain-credentials/reference/credentialaccept.md:
```diff
@@ -0,0 +1,48 @@
+| `tecDUPLICATE` | The specified credential has already been accepted. |
+| `tecEXPIRED` | The specified credential has an expiration time in the past. |
+| `tecNO_ENTRY` | The credential uniquely identified by the `Account`, `Issuer`, and `CredentialType` fields of the transaction does not exist in the ledger. |
+| `temCANNOT_PREAUTH_SELF` | The `Account` is the same as the `Issuer`. |
```
Remove this please, now you can issue credentials for yourself

oleks-rip commented on docs/xls-70d-on-chain-credentials/reference/credentialids-field.md:
```diff
@@ -0,0 +1,46 @@
+|:-----------|:------------|
+| `temDISABLED` | The related amendment is not enabled. |
+| `tecBAD_CREDENTIALS` | There was a problem with at least one of the provided credentials. For example, a specified credential doesn't belong to the sender or has not been accepted. |
+| `tecEXPIRED` | At least one provided credential has expired. |
```
Please add mention that  the expired credentials will be deleted (even if the transaction will fail with `tecEXPIRED`)

oleks-rip commented on docs/xls-70d-on-chain-credentials/reference/deposit_authorized.md:
```diff
@@ -0,0 +1,167 @@
+| `ledger_index`         | Number - [Ledger Index][] | No        | The ledger index of the ledger version that was used to generate this response. |
+| `ledger_current_index` | Number - [Ledger Index][] | No        | The ledger index of the current in-progress ledger version, which was used to generate this response. |
+| `source_account`       | String - [Address][]      | Yes       | The source account specified in the request. |
+| `validated`            | Boolean                   | No        | If `true`, the information comes from a validated ledger version. |
```
`credentials` array (if exists) added to response

oleks-rip commented on docs/xls-70d-on-chain-credentials/reference/deposit_authorized.md:
```diff
@@ -0,0 +1,167 @@
+* `actMalformed` - An [Address][] specified in the `source_account` or `destination_account` field of the request was not properly formatted. (It may contain a typo or be the wrong length, causing a failed checksum.)
+* `dstActNotFound` - The `destination_account` field of the request does not correspond to an account in the ledger.
+* `lgrNotFound` - The ledger specified by the `ledger_hash` or `ledger_index` does not exist, or it does exist but the server does not have it.
+* `srcActNotFound` - The `source_account` field of the request does not correspond to an account in the ledger.
```
`badCredentials` if credentials are expired or not accepted or doesn't exist

mvadari commented on docs/xls-70d-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+# Credentials
+
+The Credentials feature is a set of tools for managing authorization and compliance requirements using the XRP Ledger blockchain, while respecting privacy and decentralization. This feature extends and interconnects with other features of the XRP Ledger including [Decentralized Identifiers (DID)](https://xrpl.org/docs/concepts/accounts/decentralized-identifiers) and [Deposit Authorization](https: ...
```
Note: credentials don't directly connect with DIDs at all.

mDuo13 replied:
Yeah, my thought was they interconnect _indirectly_ in that you can use a DID together with a Credential for authenticating to an off-chain service, but you're right that that's vague and theoretical enough to make this statement less than helpful.

mvadari commented on docs/xls-70d-credentials/reference/credential.md:
```diff
@@ -0,0 +1,63 @@
+    "Subject": "rsUiUMpnrgxQp24dJYZDhmV4bE3aBtQyt8",
+    "Issuer": "ra5nK24KXen9AHvsdFTKHSANinZseWnPcX",
+    "CredentialType": "6D795F63726564656E7469616C",
+    "OwnerNode": "0000000000000000",
```
There are now `SubjectNode` and `IssuerNode` fields

oleks-rip commented on docs/xls-70d-credentials/reference/depositpreauth-transaction.md:
```diff
@@ -0,0 +1,91 @@
+| `tecNO_ENTRY` | The transaction tried to revoke a preauthorization that does not exist in the ledger. |
+| `tecNO_TARGET` | The transaction tried to authorize an account that is not a funded account in the ledger. |
+| `tecINSUFFICIENT_RESERVE` | The sender would not meet the [reserve requirement](https://xrpl.org/docs/concepts/accounts/reserves) after adding another entry to the ledger. (A DepositPreauth entry counts as one item towards the authorizer's owner reserve.) |
+
```
Please add tecNO_ISSUER if issuer doesn't exists

mDuo13 commented on docs/xls-70d-credentials/index.md:
```diff
@@ -0,0 +1,47 @@
+
+### Uses
+
+Within the XRP Ledger, you can use Deposit Authorization to automatically allowlist any senders with the correct set of credentials, instead of approving sending accounts individually. Future amendments could further extend credentials to gate access to other features such as holding tokens, participating in lending pools, or contributing to an AMM.
```
"Allowlist" is used (intentionally) as a verb here. It's not a code construct, so putting it as `allowlist` is incorrect. But can just use "allow" instead.

re: future amendments, I prefer it as-is because it makes it clear that those aren't features of the protocol as it exists today.

DennisDawson commented on docs/xls-70d-credentials/reference/credentialids-field.md:
```diff
@@ -0,0 +1,46 @@
+|:----------------|:-----------------|:------------------|:----------|:------------|
+| `CredentialIDs` | Array of Strings | Vector256 | No | Set of Credentials to authorize a deposit made by this transaction. Each member of the array must be the ledger entry ID of a Credential entry in the ledger and the sender of this transaction must be the subject of the credential. |
+
+You can send money to an account that uses [Deposit Authorization](https://xrpl.org/docs/concepts/accounts/depositauth) by providing this field with an exact set of credentials that are preauthorized by the recipient; the set of credentials must match a [DepositPreauth entry](./depositpreauth-entry.md) in the ledger.
```
...by the recipient. The set of credentials...

[The paragraph consists of only the one sentence, so it can easily be broken in two.]

DennisDawson commented on docs/xls-70d-credentials/reference/deposit_authorized.md:
```diff
@@ -0,0 +1,183 @@
+| Field                  | Type                      | Required? | Description   |
+|:-----------------------|:--------------------------|:----------|:--------------|
+| `credentials`          | Array of [Hash][]         | No        | The credentials specified in the request, if any. |
+| `deposit_authorized`   | Boolean                   | Yes       | Whether the specified source account is authorized to send payments directly to the destination account. If `true`, either the destination account does not require [Deposit Authorization](https://xrpl.org/docs/concepts/accounts/depositauth) or the source account is preauthorized. |
```
does not require [deposit authorization]

DennisDawson commented on docs/xls-70d-credentials/reference/deposit_authorized.md:
```diff
@@ -0,0 +1,183 @@
+* Any of the [universal error types][].
+* `invalidParams` - One or more fields are specified incorrectly, or one or more required fields are missing.
+* `actMalformed` - An [Address][] specified in the `source_account` or `destination_account` field of the request was not properly formatted. (It may contain a typo or be the wrong length, causing a failed checksum.)
+* `badCredentials` - At least one of the supplied credentials doesn't exist, is expired, or has not been accepted.
```
Just an interesting style choice that we use the contraction _doesn't_ here, but not in the next 3 lines.

DennisDawson commented on docs/xls-70d-credentials/reference/depositpreauth-entry.md:
```diff
@@ -0,0 +1,114 @@
+
+## DepositPreauth ID Format
+
+There are two formats for the ID of a `DepositPreauth` entry, depending on if it authorizes an individual account or a set of credentials.
```
...depending on whether it authorizes...

DennisDawson reviewed (approved):
Suggested optional style changes.

mDuo13 commented on docs/xls-70d-credentials/reference/credentialids-field.md:
```diff
@@ -0,0 +1,46 @@
+
+| Field           | JSON Type        | [Internal Type][] | Required? | Description |
+|:----------------|:-----------------|:------------------|:----------|:------------|
+| `CredentialIDs` | Array of Strings | Vector256 | No | Set of Credentials to authorize a deposit made by this transaction. Each member of the array must be the ledger entry ID of a Credential entry in the ledger and the sender of this transaction must be the subject of the credential. |
```
Removed the second part instead and rephrased the following paragraph about credential validity.

mDuo13 commented:
Changes in the latest commits:

- Rebased on latest `main`
- Fixed the landing page to remove dependency on legacy Redocly portal UI components and use new Cards instead.
- Addressed reviewer feedback.
- Moved the common link definitions to a snippet with the same name/path as the one in the XRPL dev portal repo, to minimize the effort of porting files over when they get incorporated into an official release.

mDuo13 commented:
Closing and reopening to work around a bug in the CI system.
